File: tiktok_shop_inventory_gate.py
"""
TikTok Shop reactive inventory gate — FINAL SAFETY ENFORCEMENT.

Ensures no invalid product attachment occurs and prevents hallucinated or missing
product IDs. Never overrides content_mode (predictive layer owns content framing).

Does not modify trends.py, trend_shift_engine.py, or existing scan pipelines.
"""

from __future__ import annotations

import json
import logging
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

_DEFAULT_STATE_PATH = Path(__file__).resolve().parent / "data" / "tiktok_shop_inventory_state.json"

_CATEGORY_RULES: list[tuple[tuple[str, ...], str]] = [
    (("army", "military", "veteran", "raf", "navy", "troops"), "military"),
    (("flag", "union jack", "union flag", "st george"), "flags"),
    (("churchill", "history", "heritage", "ww2", "d-day", "dunkirk"), "history"),
    (("hoodie", "clothing", "apparel", "wear"), "clothing"),
    (("remembrance", "poppy", "cenotaph"), "remembrance"),
    (("spitfire", "hurricane", "battle of britain"), "aviation"),
    (("king", "royal", "monarchy", "crown", "queen"), "royal"),
    (("skincare", "serum", "moisturizer", "acne"), "skincare"),
    (("makeup", "cosmetic", "lipstick", "beauty"), "beauty"),
    (("fitness", "workout", "gym", "protein"), "fitness"),
    (("book", "books"), "books"),
]


def _normalize_name(value: str) -> str:
    return re.sub(r"\s+", " ", (value or "").strip().lower())


def infer_category(product_name: str) -> str:
    """Infer a product category from the product name using keyword rules."""
    lowered = _normalize_name(product_name)
    for keywords, category in _CATEGORY_RULES:
        if any(kw in lowered for kw in keywords):
            return category
    return "general"


def _catalog_entry_name(entry: dict[str, Any]) -> str:
    return str(entry.get("name") or entry.get("product_name") or entry.get("title") or "")


def _catalog_entry_category(entry: dict[str, Any]) -> str:
    explicit = entry.get("category")
    if explicit:
        return str(explicit).lower()
    return infer_category(_catalog_entry_name(entry))


def _find_exact_match(product_name: str, catalog: list[dict[str, Any]]) -> dict[str, Any] | None:
    target = _normalize_name(product_name)
    if not target:
        return None
    for entry in catalog:
        if not isinstance(entry, dict):
            continue
        if _normalize_name(_catalog_entry_name(entry)) == target:
            return entry
    return None


def _find_category_match(product_name: str, catalog: list[dict[str, Any]]) -> dict[str, Any] | None:
    target_category = infer_category(product_name)
    if target_category == "general":
        return None
    for entry in catalog:
        if not isinstance(entry, dict):
            continue
        if _catalog_entry_category(entry) == target_category:
            product_id = entry.get("product_id")
            if product_id:
                return entry
    return None


def checkProductAvailability(
    product_name: str,
    tiktok_shop_catalog: list[dict[str, Any]],
) -> dict[str, Any]:
    """
    Search the TikTok Shop catalog for a product by exact name, then category.

    Never fabricates product IDs. Returns attachable=False when no real match exists.
    """
    catalog = [c for c in (tiktok_shop_catalog or []) if isinstance(c, dict)]
    product_name = (product_name or "").strip()

    if not product_name:
        return {
            "status": "missing",
            "product_id": None,
            "attachable": False,
            "action_required": "add_to_showcase",
            "suggested_product": product_name,
            "category": "general",
            "match_type": None,
        }

    match = _find_exact_match(product_name, catalog)
    match_type = "exact"
    if not match:
        match = _find_category_match(product_name, catalog)
        match_type = "category" if match else None

    if match:
        product_id = match.get("product_id")
        if not product_id:
            logger.warning(
                "Catalog match for %r lacks product_id — treating as missing",
                product_name,
            )
        else:
            return {
                "status": "available",
                "product_id": str(product_id),
                "attachable": True,
                "match_type": match_type,
                "matched_name": _catalog_entry_name(match),
                "category": _catalog_entry_category(match),
            }

    return {
        "status": "missing",
        "product_id": None,
        "attachable": False,
        "action_required": "add_to_showcase",
        "suggested_product": product_name,
        "category": infer_category(product_name),
        "match_type": None,
    }


def build_inventory_gap_event(availability: dict[str, Any]) -> dict[str, Any]:
    """Build the dashboard inventory_gap_event payload."""
    return {
        "product_name": availability.get("suggested_product") or "",
        "category": availability.get("category") or "general",
        "message": "Add this product to your TikTok Shop Showcase",
        "status": "waiting_user_action",
        "action_required": availability.get("action_required", "add_to_showcase"),
    }


def build_inventory_gap_notification(availability: dict[str, Any]) -> dict[str, Any]:
    """Build the pipeline inventory_gap_detected notification event."""
    product_name = availability.get("suggested_product") or "Unknown product"
    return {
        "event": "inventory_gap_detected",
        "product_name": product_name,
        "category": availability.get("category") or "general",
        "message": (
            f"Product '{product_name}' is not in your TikTok Shop catalog. "
            "Add it to Showcase to enable attachment."
        ),
        "status": "waiting_user_action",
        "action_required": "add_to_showcase",
    }


def build_blocked_attachment(
    *,
    content_id: str,
    product_name: str,
    content_mode: str,
    availability: dict[str, Any],
    reason: str,
) -> dict[str, Any]:
    """Record a reactive-layer blocked attachment (safety enforcement)."""
    return {
        "content_id": content_id,
        "product_name": product_name,
        "content_mode": content_mode,
        "reason": reason,
        "availability": availability,
        "inventory_gap_event": build_inventory_gap_event(availability),
        "notification": build_inventory_gap_notification(availability),
    }


def _load_state(path: Path | None = None) -> dict[str, Any]:
    state_path = path or _DEFAULT_STATE_PATH
    try:
        if state_path.exists():
            with open(state_path, encoding="utf-8") as f:
                data = json.load(f)
            if isinstance(data, dict):
                return data
    except Exception:
        logger.exception("Failed to load inventory state from %s", state_path)
    return {"accounts": {}}


def _save_state(state: dict[str, Any], path: Path | None = None) -> None:
    state_path = path or _DEFAULT_STATE_PATH
    state_path.parent.mkdir(parents=True, exist_ok=True)
    with open(state_path, "w", encoding="utf-8") as f:
        json.dump(state, f, indent=2)


def register_paused_attachment(
    account_id: str,
    *,
    product_name: str,
    content_id: str,
    pipeline_run_id: str,
    availability: dict[str, Any],
    state_path: Path | None = None,
) -> dict[str, Any]:
    """Record a paused product attachment for later resume."""
    state = _load_state(state_path)
    accounts = state.setdefault("accounts", {})
    account = accounts.setdefault(account_id, {"paused_attachments": []})

    paused = {
        "content_id": content_id,
        "pipeline_run_id": pipeline_run_id,
        "product_name": product_name,
        "category": availability.get("category"),
        "inventory_gap_event": build_inventory_gap_event(availability),
        "notification": build_inventory_gap_notification(availability),
        "paused_at": datetime.now(timezone.utc).isoformat(),
        "status": "waiting_user_action",
    }
    account["paused_attachments"] = [
        p for p in account.get("paused_attachments", [])
        if p.get("content_id") != content_id
    ]
    account["paused_attachments"].append(paused)
    _save_state(state, state_path)
    return paused


def get_paused_attachments(
    account_id: str,
    state_path: Path | None = None,
) -> list[dict[str, Any]]:
    state = _load_state(state_path)
    account = state.get("accounts", {}).get(account_id, {})
    return list(account.get("paused_attachments", []))


def resumeAfterInventoryUpdate(
    account_id: str,
    tiktok_shop_catalog: list[dict[str, Any]],
    *,
    content_id: str | None = None,
    state_path: Path | None = None,
) -> dict[str, Any]:
    """
    Re-run product match ONLY for paused attachments — does not re-run the full pipeline.

    If the product is now available, marks the attachment as resumed with the real product_id.
  Never modifies content_mode.
    """
    state = _load_state(state_path)
    account = state.get("accounts", {}).get(account_id)
    if not account:
        return {
            "success": True,
            "account_id": account_id,
            "resumed": [],
            "still_waiting": [],
            "message": "No paused attachments for this account.",
        }

    paused_list = account.get("paused_attachments", [])
    if content_id:
        paused_list = [p for p in paused_list if p.get("content_id") == content_id]

    resumed: list[dict[str, Any]] = []
    still_waiting: list[dict[str, Any]] = []

    for paused in paused_list:
        product_name = paused.get("product_name") or ""
        availability = checkProductAvailability(product_name, tiktok_shop_catalog)

        if availability.get("attachable"):
            attachment = {
                "content_id": paused.get("content_id"),
                "pipeline_run_id": paused.get("pipeline_run_id"),
                "product_name": product_name,
                "product_id": availability["product_id"],
                "status": "attached",
                "resumed_at": datetime.now(timezone.utc).isoformat(),
                "match_type": availability.get("match_type"),
            }
            resumed.append(attachment)
            paused["status"] = "resumed"
            paused["product_id"] = availability["product_id"]
        else:
            still_waiting.append({
                **paused,
                "inventory_gap_event": build_inventory_gap_event(availability),
                "status": "waiting_user_action",
            })

    remaining = [
        p for p in account.get("paused_attachments", [])
        if p.get("status") != "resumed"
    ]
    account["paused_attachments"] = remaining
    _save_state(state, state_path)

    return {
        "success": True,
        "account_id": account_id,
        "resumed": resumed,
        "still_waiting": still_waiting,
        "message": (
            f"Resumed {len(resumed)} attachment(s); "
            f"{len(still_waiting)} still waiting for Showcase onboarding."
        ),
    }


__all__ = [
    "infer_category",
    "checkProductAvailability",
    "build_inventory_gap_event",
    "build_inventory_gap_notification",
    "build_blocked_attachment",
    "register_paused_attachment",
    "get_paused_attachments",
    "resumeAfterInventoryUpdate",
]

File: dashboard-sync/tiktok-shop-inventory-gate.js
/**
 * TikTok Shop Inventory Gate — human-in-the-loop product onboarding.
 *
 * Detects missing catalog products, pauses ONLY product attachment, and resumes
 * after the creator adds items to TikTok Shop Showcase.
 *
 * Isolated from existing trend UI and niche comment intelligence modules.
 */
(function () {
  "use strict";

  var MOUNT_ID = "tiktokShopInventoryGate";
  var CATALOG_KEY = "tiktok_shop_catalog";
  var PAUSED_KEY = "tiktok_shop_paused_attachments";
  var OVERRIDE_KEY = "tiktok_shop_inventory_override";

  var CATEGORY_RULES = [
    { keywords: ["army", "military", "veteran", "raf", "navy", "troops"], category: "military" },
    { keywords: ["flag", "union jack", "union flag", "st george"], category: "flags" },
    { keywords: ["churchill", "history", "heritage", "ww2", "d-day", "dunkirk"], category: "history" },
    { keywords: ["hoodie", "clothing", "apparel", "wear"], category: "clothing" },
    { keywords: ["remembrance", "poppy", "cenotaph"], category: "remembrance" },
    { keywords: ["spitfire", "hurricane", "battle of britain"], category: "aviation" },
    { keywords: ["king", "royal", "monarchy", "crown", "queen"], category: "royal" },
    { keywords: ["skincare", "serum", "moisturizer", "acne"], category: "skincare" },
    { keywords: ["makeup", "cosmetic", "lipstick", "beauty"], category: "beauty" },
    { keywords: ["fitness", "workout", "gym", "protein"], category: "fitness" },
    { keywords: ["book", "books"], category: "books" }
  ];

  function normalizeName(value) {
    return String(value || "").trim().toLowerCase().replace(/\s+/g, " ");
  }

  function inferCategory(productName) {
    var lowered = normalizeName(productName);
    for (var i = 0; i < CATEGORY_RULES.length; i++) {
      var rule = CATEGORY_RULES[i];
      for (var k = 0; k < rule.keywords.length; k++) {
        if (lowered.indexOf(rule.keywords[k]) !== -1) return rule.category;
      }
    }
    return "general";
  }

  function catalogEntryName(entry) {
    return String(entry.name || entry.product_name || entry.title || "");
  }

  function catalogEntryCategory(entry) {
    if (entry.category) return String(entry.category).toLowerCase();
    return inferCategory(catalogEntryName(entry));
  }

  function findExactMatch(productName, catalog) {
    var target = normalizeName(productName);
    if (!target) return null;
    for (var i = 0; i < catalog.length; i++) {
      if (normalizeName(catalogEntryName(catalog[i])) === target) return catalog[i];
    }
    return null;
  }

  function findCategoryMatch(productName, catalog) {
    var targetCategory = inferCategory(productName);
    if (targetCategory === "general") return null;
    for (var i = 0; i < catalog.length; i++) {
      var entry = catalog[i];
      if (catalogEntryCategory(entry) === targetCategory && entry.product_id) return entry;
    }
    return null;
  }

  function checkProductAvailability(productName, tiktokShopCatalog) {
    var catalog = (tiktokShopCatalog || []).filter(function (c) { return c && typeof c === "object"; });
    productName = String(productName || "").trim();

    if (!productName) {
      return {
        status: "missing",
        product_id: null,
        attachable: false,
        action_required: "add_to_showcase",
        suggested_product: productName,
        category: "general",
        match_type: null
      };
    }

    var match = findExactMatch(productName, catalog);
    var matchType = "exact";
    if (!match) {
      match = findCategoryMatch(productName, catalog);
      matchType = match ? "category" : null;
    }

    if (match && match.product_id) {
      return {
        status: "available",
        product_id: String(match.product_id),
        attachable: true,
        match_type: matchType,
        matched_name: catalogEntryName(match),
        category: catalogEntryCategory(match)
      };
    }

    return {
      status: "missing",
      product_id: null,
      attachable: false,
      action_required: "add_to_showcase",
      suggested_product: productName,
      category: inferCategory(productName),
      match_type: null
    };
  }

  function buildInventoryGapEvent(availability) {
    return {
      product_name: availability.suggested_product || "",
      category: availability.category || "general",
      message: "Add this product to your TikTok Shop Showcase",
      status: "waiting_user_action",
      action_required: availability.action_required || "add_to_showcase"
    };
  }

  function getCatalog() {
    try {
      var raw = localStorage.getItem(CATALOG_KEY);
      if (!raw) return [];
      var parsed = JSON.parse(raw);
      return Array.isArray(parsed) ? parsed : [];
    } catch (e) {
      return [];
    }
  }

  function saveCatalog(catalog) {
    localStorage.setItem(CATALOG_KEY, JSON.stringify(catalog));
  }

  function getPausedAttachments() {
    try {
      var raw = localStorage.getItem(PAUSED_KEY);
      if (!raw) return [];
      var parsed = JSON.parse(raw);
      return Array.isArray(parsed) ? parsed : [];
    } catch (e) {
      return [];
    }
  }

  function savePausedAttachments(items) {
    localStorage.setItem(PAUSED_KEY, JSON.stringify(items));
  }

  function getAccountId() {
    if (typeof currentUser !== "undefined" && currentUser && currentUser.id) {
      return currentUser.id;
    }
    return "local_account";
  }

  function emitInventoryGapDetected(availability, metadata) {
    var gapEvent = buildInventoryGapEvent(availability);
    if (typeof trackEvent === "function") {
      trackEvent("inventory_gap_detected", {
        inventory_gap_event: gapEvent,
        product_name: gapEvent.product_name,
        category: gapEvent.category,
        status: gapEvent.status,
        metadata: metadata || {}
      });
    }
    return gapEvent;
  }

  function suggestProductName(keyword) {
    if (typeof makeProduct === "function") return makeProduct(keyword);
    return keyword ? String(keyword) + " product" : "";
  }

  function registerPausedAttachment(contentId, productName, availability, keyword) {
    var paused = getPausedAttachments();
    var record = {
      content_id: contentId,
      account_id: getAccountId(),
      product_name: productName,
      keyword: keyword || "",
      category: availability.category,
      inventory_gap_event: buildInventoryGapEvent(availability),
      status: "waiting_user_action",
      paused_at: new Date().toISOString()
    };
    paused = paused.filter(function (p) { return p.content_id !== contentId; });
    paused.push(record);
    savePausedAttachments(paused);
    return record;
  }

  function resumeAfterInventoryUpdate(accountId, contentId) {
    var catalog = getCatalog();
    var paused = getPausedAttachments();
    var targetAccount = accountId || getAccountId();
    var resumed = [];
    var stillWaiting = [];

    for (var i = 0; i < paused.length; i++) {
      var item = paused[i];
      if (item.account_id !== targetAccount) continue;
      if (contentId && item.content_id !== contentId) continue;

      var availability = checkProductAvailability(item.product_name, catalog);
      if (availability.attachable) {
        resumed.push({
          content_id: item.content_id,
          product_name: item.product_name,
          product_id: availability.product_id,
          status: "attached",
          resumed_at: new Date().toISOString()
        });
        item.status = "resumed";
        item.product_id = availability.product_id;
      } else {
        stillWaiting.push({
          content_id: item.content_id,
          product_name: item.product_name,
          inventory_gap_event: buildInventoryGapEvent(availability),
          status: "waiting_user_action"
        });
      }
    }

    var remaining = paused.filter(function (p) {
      return p.status !== "resumed";
    });
    savePausedAttachments(remaining);

    if (typeof trackEvent === "function") {
      trackEvent("inventory_gap_resume", {
        account_id: targetAccount,
        resumed_count: resumed.length,
        still_waiting_count: stillWaiting.length
      });
    }

    return {
      success: true,
      account_id: targetAccount,
      resumed: resumed,
      still_waiting: stillWaiting
    };
  }

  function gateProductAttachment(keyword, productName, contentId) {
    var catalog = getCatalog();
    var resolvedName = productName || suggestProductName(keyword);
    var availability = checkProductAvailability(resolvedName, catalog);
    var id = contentId || "content_" + normalizeName(keyword).replace(/\s+/g, "_");

    if (availability.attachable) {
      return {
        content_id: id,
        keyword: keyword,
        product_name: resolvedName,
        attachment_status: "attached",
        product_id: availability.product_id,
        paused: false
      };
    }

    var paused = registerPausedAttachment(id, resolvedName, availability, keyword);
    var gapEvent = emitInventoryGapDetected(availability, { content_id: id, keyword: keyword });

    return {
      content_id: id,
      keyword: keyword,
      product_name: resolvedName,
      attachment_status: "paused_inventory_gap",
      product_id: null,
      paused: true,
      inventory_gap_event: gapEvent,
      paused_record: paused
    };
  }

  function renderInventoryGapCard(gap) {
    var h = '<div class="card" style="border-color:rgba(255,71,87,.35);margin-bottom:12px">';
    h += '<div class="card-header"><h2>TikTok Shop Inventory Gap</h2>';
    h += '<div class="section-icon" style="color:#ff4757">&#9888;</div></div>';
    h += '<p style="font-size:12px;color:var(--muted);margin-bottom:10px">';
    h += 'Product attachment is paused until this item is in your TikTok Shop Showcase. ';
    h += 'Your content plan and pipeline continue normally.';
    h += '</p>';
    h += '<div style="padding:12px;background:rgba(255,71,87,.06);border:1px solid rgba(255,71,87,.2);border-radius:10px;margin-bottom:10px">';
    h += '<div style="font-size:10px;color:#ff4757;font-weight:800;text-transform:uppercase;letter-spacing:1px;margin-bottom:4px">Missing Product</div>';
    h += '<div style="font-size:16px;font-weight:900;margin-bottom:4px">' + escapeHtml(gap.product_name) + '</div>';
    h += '<div style="font-size:10px;color:var(--muted)">Category: ' + escapeHtml(gap.category || "general") + '</div>';
    h += '<div style="font-size:11px;color:var(--white);margin-top:8px">' + escapeHtml(gap.message || "Add this product to your TikTok Shop Showcase") + '</div>';
    h += '</div>';
    h += '<div style="display:flex;gap:8px;flex-wrap:wrap">';
    h += '<button class="copy-btn" style="flex:1;background:rgba(255,71,87,.12);color:#ff4757;border-color:rgba(255,71,87,.35)" ';
    h += 'onclick="TikTokShopInventoryGate.handleAddToShowcase(\'' + escapeAttr(gap.content_id) + '\')">';
    h += 'Add to Showcase</button>';
    h += '<button class="copy-btn" style="flex:1" ';
    h += 'onclick="TikTokShopInventoryGate.handleUserOverride(\'' + escapeAttr(gap.content_id) + '\')">';
    h += 'Continue Without Product</button>';
    h += '</div>';
    h += '<p style="font-size:9px;color:var(--muted);margin-top:8px;text-align:center">';
    h += 'After adding in TikTok Shop, click "Add to Showcase" to re-check your catalog and resume attachment.';
    h += '</p></div>';
    return h;
  }

  function escapeHtml(value) {
    return String(value || "")
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;")
      .replace(/"/g, "&quot;");
  }

  function escapeAttr(value) {
    return String(value || "").replace(/'/g, "\\'");
  }

  function render(catalogGaps) {
    var el = document.getElementById(MOUNT_ID);
    if (!el) return;

    var gaps = catalogGaps || [];
    var paused = getPausedAttachments().filter(function (p) {
      return p.status === "waiting_user_action";
    });

    if (gaps.length === 0 && paused.length === 0) {
      el.innerHTML = "";
      return;
    }

    var seen = {};
    var h = "";
    var allGaps = gaps.concat(paused.map(function (p) {
      return {
        content_id: p.content_id,
        product_name: p.product_name,
        category: p.category,
        message: (p.inventory_gap_event && p.inventory_gap_event.message) || "Add this product to your TikTok Shop Showcase",
        status: "waiting_user_action"
      };
    }));

    for (var i = 0; i < allGaps.length; i++) {
      var gap = allGaps[i];
      var key = gap.content_id || gap.product_name;
      if (seen[key]) continue;
      seen[key] = true;
      h += renderInventoryGapCard(gap);
    }

    el.innerHTML = h;
  }

  function processTrendResults(results) {
    if (!results || !results.length) {
      render([]);
      return { gaps: [], attachments: [] };
    }

    var catalog = getCatalog();
    if (!catalog.length) {
      seedDefaultCatalog();
      catalog = getCatalog();
    }

    var productPick = results.length >= 3 ? results[2] : results[results.length - 1];
    var keyword = productPick.keyword || "";
    var productName = productPick.product || suggestProductName(keyword);
    var attachment = gateProductAttachment(keyword, productName, "plan_slot_3");

    var gaps = [];
    var attachments = [];

    if (attachment.paused) {
      gaps.push({
        content_id: attachment.content_id,
        product_name: attachment.product_name,
        category: attachment.inventory_gap_event.category,
        message: attachment.inventory_gap_event.message,
        status: "waiting_user_action"
      });
    } else {
      attachments.push(attachment);
    }

    render(gaps);
    return { gaps: gaps, attachments: attachments };
  }

  function seedDefaultCatalog() {
    if (getCatalog().length > 0) return;
    saveCatalog([
      { product_id: "tts_10001", name: "British Army history books", category: "military" },
      { product_id: "tts_10002", name: "Union Jack flags and patriotic decor", category: "flags" },
      { product_id: "tts_10003", name: "Royal family collectibles", category: "royal" },
      { product_id: "tts_10004", name: "British history books", category: "history" },
      { product_id: "tts_10005", name: "Proudly British merchandise", category: "general" }
    ]);
  }

  function handleAddToShowcase(contentId) {
    var result = resumeAfterInventoryUpdate(getAccountId(), contentId);
    if (result.resumed.length > 0) {
      if (typeof showToast === "function") {
        showToast("Product found in catalog — attachment resumed!");
      }
      render([]);
      if (typeof renderAllSections === "function" && typeof cachedResults !== "undefined" && cachedResults) {
        processTrendResults(cachedResults);
      }
    } else if (result.still_waiting.length > 0) {
      if (typeof showToast === "function") {
        showToast("Still not in catalog — add the product to Showcase, then try again.");
      }
      render(result.still_waiting.map(function (w) {
        return {
          content_id: w.content_id,
          product_name: w.product_name,
          category: (w.inventory_gap_event && w.inventory_gap_event.category) || "general",
          message: (w.inventory_gap_event && w.inventory_gap_event.message) || "Add this product to your TikTok Shop Showcase",
          status: "waiting_user_action"
        };
      }));
    }
  }

  function handleUserOverride(contentId) {
    localStorage.setItem(OVERRIDE_KEY + "_" + contentId, "1");
    var paused = getPausedAttachments().filter(function (p) {
      return p.content_id !== contentId;
    });
    savePausedAttachments(paused);
    if (typeof trackEvent === "function") {
      trackEvent("inventory_gap_user_override", { content_id: contentId });
    }
    if (typeof showToast === "function") {
      showToast("Continuing without product attachment for this slot.");
    }
    render([]);
  }

  function init() {
    seedDefaultCatalog();
    var el = document.getElementById(MOUNT_ID);
    if (!el) return;
    var paused = getPausedAttachments().filter(function (p) {
      return p.status === "waiting_user_action";
    });
    if (paused.length) {
      render(paused.map(function (p) {
        return {
          content_id: p.content_id,
          product_name: p.product_name,
          category: p.category,
          message: (p.inventory_gap_event && p.inventory_gap_event.message) || "Add this product to your TikTok Shop Showcase",
          status: "waiting_user_action"
        };
      }));
    }
  }

  window.TikTokShopInventoryGate = {
    checkProductAvailability: checkProductAvailability,
    resumeAfterInventoryUpdate: resumeAfterInventoryUpdate,
    gateProductAttachment: gateProductAttachment,
    processTrendResults: processTrendResults,
    handleAddToShowcase: handleAddToShowcase,
    handleUserOverride: handleUserOverride,
    getCatalog: getCatalog,
    saveCatalog: saveCatalog,
    render: render,
    init: init
  };

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
})();
